### PR TITLE
[FEATURE] Improve bootstrap quantile method accuracy

### PR DESCRIPTION
**Pull request by NathanFarmer**
Changes proposed in this pull request:
- GREAT-499
- Add bias correction to bootstrap estimation method
- Test performance improvement by comparing against population parameters as opposed to sample statistics
- Document changes and clarify legacy methods that have been tested
- Investigation of replacing this method with `scipy.stats.bootstrap`

### Definition of Done
- [X] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [X] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [X] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: tests/rule_based_profiler/parameter_builder/test_simple_date_format_string_parameter_builder.py [hunk 1]
```diff
@@ -13,6 +13,9 @@
     ParameterContainer,
     get_parameter_value_by_fully_qualified_parameter_name,
 )
+from great_expectations.rule_based_profiler.util import (
+    get_parameter_value_and_validate_return_type,
+)
```
Comment (NathanFarmer) [inline tests/rule_based_profiler/parameter_builder/test_simple_date_format_string_parameter_builder.py]: All changes in `test_simple_date_format_string_parameter_builder` are an unrelated refactor to add type checking to previous work.

File: great_expectations/rule_based_profiler/util.py [hunk 2]
```diff
@@ -242,36 +242,80 @@ def compute_quantiles(
         metric_values,
         q=(false_positive_rate / 2),
         axis=0,
-        interpolation="linear",  # can be omitted ("linear" is default)
```
Comment (NathanFarmer) [inline great_expectations/rule_based_profiler/util.py]: `interpolation` is deprecated in favor of `method`. This means using `interpolation` emits warnings and using `method` causes errors in Python 3.6 and 3.7. Since we are using the default method I have chosen to remove this parameter.

File: great_expectations/rule_based_profiler/util.py [hunk 3]
```diff
@@ -242,36 +242,80 @@ def compute_quantiles(
...
-    Internal implementation of the "bootstrap" estimator method, returning confidence interval for a distribution.
-    See https://en.wikipedia.org/wiki/Bootstrapping_(statistics) for an introduction to "bootstrapping" in statistics.
+    An internal implementation of the "bootstrap" estimator method, returning a point estimate for a population
+    parameter of interest (lower and upper quantiles in this case). See
+    https://en.wikipedia.org/wiki/Bootstrapping_(statistics) for an introduction to "bootstrapping" in statistics.
+
+    This implementation has been replaced by "compute_bootstrap_quantiles_bias_corrected_point_estimate" and only
+    remains to demonstrate the performance improvement achieved by correcting for bias. Upon the implementation of a
+    Machine Learning Lifecycle framework, the performance improvement can be documented and this legacy method can be
+    removed from the codebase.
```
Comment (NathanFarmer) [inline great_expectations/rule_based_profiler/util.py]: Justification for leaving in the legacy method.

File: tests/conftest.py [hunk 4]
```diff
@@ -6476,7 +6476,7 @@ def bobster_columnar_table_multi_batch_normal_mean_5000_stdev_1000():
         1.0e3
     )
     my_row_count_range_rule_expect_table_row_count_to_be_between_expectation_num_stds: float = (
-        3.00
+        4.00
```
Comment (NathanFarmer) [inline tests/conftest.py]: This test was failing intermittently because 3 standard deviations only covers 99.7% of the normal distribution.

File: great_expectations/rule_based_profiler/util.py (hunk 3 above)
Comment (Shinnnyshinshin) [reply great_expectations/rule_based_profiler/util.py]: Wow, really well-cited and explained. Thank very much @NathanFarmer

File: great_expectations/rule_based_profiler/util.py [hunk 5]
```diff
@@ -242,36 +242,80 @@ def compute_quantiles(
...
+    return lower_quantile_point_estimate, upper_quantile_point_estimate
 
-    This implementation is sub-par compared to the one available from the "SciPy" standard library
-    ("https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.bootstrap.html"), because it introduces bias and
-    does not handle multi-dimensional statistics (unlike "scipy.stats.bootstrap", which corrects for bias and is
-    vectorized, thus having the ability to accept a multi-dimensional statistic function and process all dimensions).
 
-    This implementation will be replaced by "scipy.stats.bootstrap" when Great Expectations can be upgraded to use a
-    more up-to-date version of the "scipy" Python package (the currently used version does not have "bootstrap").
+def compute_bootstrap_quantiles_bias_corrected_point_estimate(
```
Comment (Shinnnyshinshin) [inline great_expectations/rule_based_profiler/util.py]: :heart: Thank you @NathanFarmer for digging into this. I think this also sets the bar for most-extensively researched and best-supported DocString we have in our codebase (by far)

**Review by Shinnnyshinshin — APPROVED** [review]
Thank you @NathanFarmer. I'm approving, but asking for at least one more approval.. just we can also enjoy this really well documented and written PR :)

File: great_expectations/rule_based_profiler/util.py [hunk 6]
```diff
@@ -299,20 +343,37 @@ def compute_bootstrap_quantiles(
...
         metric_values, size=(n_resamples, metric_values.size)
     )
-    lower_quantile = np.mean(
-        np.quantile(
-            bootstraps,
-            q=false_positive_rate / 2,
-            axis=1,
-            interpolation="linear",  # can be omitted ("linear" is default)
-        )
+    lower_quantiles: Union[np.nd_array, Number] = np.quantile(
```
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/util.py]: @NathanFarmer What is `np.nd_array`?  I am aware of only `np.ndarray` -- could this be a typo?  Also, please note that we already have the types `MetricValue` (singular) and `MetricValues` (plural), which encompass the needed types.  I am also not sure whether or not we need multi-dimensional statistics right now -- we could use the same statistic computed for every dimension, though (is that what you meant)?  Happy to discuss.  Thank you.

File: great_expectations/rule_based_profiler/util.py [hunk 7]
```diff
@@ -299,20 +343,37 @@ def compute_bootstrap_quantiles(
...
+        axis=1,
+    )
+    upper_quantile_point_estimate: Number = np.mean(upper_quantiles)
+
+    upper_quantile_residuals: Union[np.ndarray, Number] = (
+        upper_quantiles - upper_quantile_point_estimate
+    )
+    upper_quantile_bias: Number = sum(upper_quantile_residuals) / len(upper_quantiles)
+    upper_quantile_bias_corrected_point_estimate: Number = (
+        upper_quantile_point_estimate + upper_quantile_bias
```
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/util.py]: @NathanFarmer Where is this bias correction formula described?  Could you please point me to a resource where I can see the proof of the math involved?  Essentially, the code says (where "^T" denotes vector/matrix transpose):
```
 mean(statistic) + [statistic - mean(statistic)]^T * [1-vector] / N
```
which is to say "mean of statistic, corrected by average deviation between statistic and mean of statistic itself").  I would be curious to see the math the shows the reason that this bias correction method is advantageous.  Thank you!

Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/util.py]: @NathanFarmer I do recall using a different approach.  It may well have been using the quantiles (upper, lower) as the statistics and returning the arithmetic mean of the confidence interval, but I am not sure -- still looking.  I also recall using the `SciPy` bootstrap API to compute all the bootstraps for us by supplying it with the data.  I feel that we should not be doing this ourselves (i.e., we should not be computing the bootstraps ourselves, we should be leveraging the SciPy library for this).  Thanks.

File: tests/rule_based_profiler/parameter_builder/test_parameter_computations.py [hunk 8]
```diff
@@ -29,27 +32,136 @@ def _generate_distribution_samples(size: Optional[int] = 36) -> pd.DataFrame:
...
+
+
+def test_bootstrap_bias_corrected_point_estimate_efficacy():
+    df: pd.DataFrame = _generate_distribution_samples(size=1000)
+    false_positive_rate: np.float64 = np.float64(0.01)
+    columns: pd.Index = df.columns
+    column: str
+    lower_quantile_bias_corrected_point_estimate: np.float64
+    upper_quantile_bias_corrected_point_estimate: np.float64
+    actual_false_positive_rates: Dict[str, Union[float, np.float64]] = {}
```
Comment (alexsherstinsky) [inline tests/rule_based_profiler/parameter_builder/test_parameter_computations.py:178]: @NathanFarmer Shall we use the `MetricValue` / `MetricValues` types we defined, or leave it as is?

File: tests/rule_based_profiler/parameter_builder/test_parameter_computations.py [hunk 9]
```diff
@@ -29,27 +32,136 @@ def _generate_distribution_samples(size: Optional[int] = 36) -> pd.DataFrame:
...
+                / df.shape[0]
+            )
+
+            improvement.append(
+                actual_biased_false_positive_rates[column]
+                >= actual_bias_corrected_false_positive_rates[column]
+            )
+
+    # bias correction should consistently improve performance at least 90% of the time
+    assert sum(improvement) / len(improvement) >= 0.90
```
Comment (alexsherstinsky) [inline tests/rule_based_profiler/parameter_builder/test_parameter_computations.py]: @NathanFarmer Very interesting result.  And we do not get an improvement by using the arithmetic means of the quantiles statistics based confidence intervals?  I am trying to find what I did, which also showed an improvement over the "biased" estimate.  Have we compared the said confidence intervals based approach with the implemented bias-correction based technique (while letting the bootstraps for the confidence intervals based approach be computed as part of `SciPy.bootstra()` method call?  Thank you.

File: tests/rule_based_profiler/parameter_builder/test_simple_date_format_string_parameter_builder.py [hunk 10]
```diff
@@ -11,10 +11,9 @@
 from great_expectations.rule_based_profiler.parameter_builder.simple_date_format_string_parameter_builder import (
     DEFAULT_CANDIDATE_STRINGS,
 )
-from great_expectations.rule_based_profiler.types import (
-    Domain,
-    ParameterContainer,
-    get_parameter_value_by_fully_qualified_parameter_name,
+from great_expectations.rule_based_profiler.types import Domain, ParameterContainer
```
Comment (alexsherstinsky) [inline tests/rule_based_profiler/parameter_builder/test_simple_date_format_string_parameter_builder.py:17]: The existing code worked:
```
from great_expectations.rule_based_profiler.types import (
    Domain,
    ParameterContainer,
    get_parameter_value_by_fully_qualified_parameter_name,
)
```
because all (including the method in the last line) were defined in the `types` module.  Is there a reason to change?  Thanks.

File: tests/integration/profiling/rule_based_profilers/test_profiler_user_workflows.py [hunk 11]
```diff
@@ -935,6 +935,7 @@ def test_bobby_expect_column_values_to_be_between_auto_yes_default_profiler_conf
     version.parse(np.version.version) < version.parse("1.21.0"),
     reason="requires numpy version 1.21.0 or newer",
 )
+@probabilistic_test
```
Comment (alexsherstinsky) [inline tests/integration/profiling/rule_based_profilers/test_profiler_user_workflows.py]: @NathanFarmer We do not need a second decorator call to `@probabilistic_test` -- there is one already in line 933 -- thanks!

**Review by alexsherstinsky — CHANGES_REQUESTED** [review]
@NathanFarmer I found some typos and also made suggestions (please do not feel compelled to agree with them).  The main questions are:
* the math concerning the bias correction;
* use of the `SciPy` library to generate the `bootstraps`

I finally found the algorithm I implemented in June/July-2021 (which used `SciPy.stats.bootstrap()`.  Here it is for your consideration:

```
    def _get_bootstrap_confidence_interval(
        self,
        metric_values: Union[np.ndarray, List[Number]],
        confidence_level: np.float64,
        domain: Domain,
        *,
        variables: Optional[ParameterContainer] = None,
        parameters: Optional[Dict[str, ParameterContainer]] = None,
    ):
        # Obtain num_bootstrap_samples override from rule state (i.e., variables and parameters); from instance variable otherwise.
        num_bootstrap_samples: Optional[
            int
        ] = get_parameter_value_and_validate_return_type(
            domain=domain,
            parameter_reference=self._num_bootstrap_samples,
            expected_return_type=None,
            variables=variables,
            parameters=parameters,
        )

        rng: np.random.Generator = np.random.default_rng()

        bootstrap_samples: tuple = (
            metric_values,
        )  # bootstrap samples must be in a sequence

        # noinspection PyPep8Naming
        BootstrapResult = make_dataclass(
            "BootstrapResult", ["confidence_interval", "standard_error"]
        )

        bootstrap_result_low: BootstrapResult
        bootstrap_result_high: BootstrapResult

        uncertainty_level: float = 1.0 - confidence_level
        quantile_probability_lower: float = 5.0e-1 * uncertainty_level
        quantile_probability_upper: float = 1.0 - quantile_probability_lower

        if num_bootstrap_samples is None:
            bootstrap_result_low = bootstrap(
                bootstrap_samples,
                lambda data: np.quantile(
                    data,
                    q=quantile_probability_lower,
                ),
                vectorized=False,
                confidence_level=confidence_level,
                random_state=rng,
            )
            bootstrap_result_high = bootstrap(
                bootstrap_samples,
                lambda data: np.quantile(
                    data,
                    q=quantile_probability_upper,
                ),
                vectorized=False,
                confidence_level=confidence_level,
                random_state=rng,
            )
        else:
            bootstrap_result_low = bootstrap(
                bootstrap_samples,
                lambda data: np.quantile(
                    data,
                    q=quantile_probability_lower,
                ),
                vectorized=False,
                confidence_level=confidence_level,
                n_resamples=num_bootstrap_samples,
                random_state=rng,
            )
            bootstrap_result_high = bootstrap(
                bootstrap_samples,
                lambda data: np.quantile(
                    data,
                    q=quantile_probability_upper,
                ),
                vectorized=False,
                confidence_level=confidence_level,
                n_resamples=num_bootstrap_samples,
                random_state=rng,
            )

        confidence_interval_low: ConfidenceInterval = (
            bootstrap_result_low.confidence_interval
        )
        confidence_interval_low_mean: np.float64 = np.mean(
            [confidence_interval_low.low, confidence_interval_low.high]
        )
        confidence_interval_high: ConfidenceInterval = (
            bootstrap_result_high.confidence_interval
        )
        confidence_interval_high_mean: np.float64 = np.mean(
            [confidence_interval_high.low, confidence_interval_high.high]
        )

        return ConfidenceInterval(
            low=confidence_interval_low_mean, high=confidence_interval_high_mean
        )
```

Aside from the type hints that are not current, compared to what we use now, here are the key points:
* Back then, the random number generated needed to be supplied (not sure about now).
* We used the confidence-level style data classes as types (to match what the `SciPy.stats.bootstrap()` library's mechanism uses).
* Crucially, we defined two curried `np.quantile` statistics (corresponding to `quantile_probability_lower` and `quantile_probability_upper`) -- being able to use curried functions as statistics is one of the major advantages of the `SciPy.stats` library that I am hoping we can leverage.
* Differentiating between the default and customized `n_samples` is a little awkward, but that is how it was back then -- perhaps it is better now (worth taking a look; not a big deal).
* In the end, we return the means of the confidence intervals for the low and high ends of the range.

The overall results seemed consistently better than that legacy.  It would be really interesting to compare them to your bias-corrected estimates, and, whichever is better, I would still hope that we can find a way to leverage the bootstraps generating and statistic function currying capabilities of the standard library.

Thank you very much!

File: tests/rule_based_profiler/parameter_builder/test_simple_date_format_string_parameter_builder.py (hunk 10 above)
Comment (NathanFarmer) [reply tests/rule_based_profiler/parameter_builder/test_simple_date_format_string_parameter_builder.py:17]: This change is due to the added benefit of validating the return type in `get_parameter_value_and_validate_return_type`. In recent PRs we have been using this method in favor of the removed method in cases where we know what return type to expect.

File: tests/integration/profiling/rule_based_profilers/test_profiler_user_workflows.py (hunk 11 above)
Comment (NathanFarmer) [reply tests/integration/profiling/rule_based_profilers/test_profiler_user_workflows.py]: Nice catch! I didn't notice that line 933 got merged in from develop.

File: great_expectations/rule_based_profiler/util.py (hunk 6 above)
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/util.py]: Yes, `np.nd_array` was a typo (it has since been corrected to `np.ndarray`). I'm not sure if `MetricValues` applies here since this is the bootstrap point estimate built using the `MetricValue` for each batch.

File: great_expectations/rule_based_profiler/util.py (hunk 7 above)
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/util.py]: This is outdated and this method has been changed. See:
```
Efron, B., & Tibshirani, R. J. (1993). Estimates of bias. An Introduction to the Bootstrap (pp. 124-130).
    Springer Science and Business Media Dordrecht. DOI 10.1007/978-1-4899-4541-9
```
for the new method.

File: tests/rule_based_profiler/parameter_builder/test_parameter_computations.py (hunk 8 above)
Comment (NathanFarmer) [reply tests/rule_based_profiler/parameter_builder/test_parameter_computations.py:178]: This is the target one-sided quantile for which to estimate the `MetricValue`. I feel like it should not be considered a `MetricValue`.

File: tests/rule_based_profiler/parameter_builder/test_parameter_computations.py (hunk 9 above)
Comment (NathanFarmer) [reply tests/rule_based_profiler/parameter_builder/test_parameter_computations.py]: This is outdated. See new performance measurement methods.

File: great_expectations/rule_based_profiler/helpers/util.py [hunk 12]
```diff
@@ -243,36 +244,93 @@ def compute_quantiles(
...
     return lower_quantile, upper_quantile
 
 
-def compute_bootstrap_quantiles(
+def compute_bootstrap_quantiles_point_estimate(
     metric_values: np.ndarray,
     false_positive_rate: np.float64,
     n_resamples: int,
 ) -> Tuple[Number, Number]:
+    # This method is the winner of our performance testing
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/helpers/util.py]: What were the candidates? Should we possibly include that context here or is it not important?

Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/helpers/util.py]: Good call. I have documented the candidate methods here as well now.

File: tests/rule_based_profiler/parameter_builder/test_simple_date_format_string_parameter_builder.py (hunk 10 above)
Comment (alexsherstinsky) [reply tests/rule_based_profiler/parameter_builder/test_simple_date_format_string_parameter_builder.py:17]: @NathanFarmer Probably GitHub highlighting confused me.  I will check again.  Thanks!

File: great_expectations/rule_based_profiler/helpers/util.py [hunk 13]
```diff
@@ -297,23 +358,148 @@ def compute_bootstrap_quantiles(
...
-            axis=1,
-            interpolation="linear",  # can be omitted ("linear" is default)
+
+    bootstrap_lower_quantiles: Union[np.ndarray, Number] = np.quantile(
+        bootstraps,
+        q=lower_quantile_pct,
+        axis=1,
+    )
+    bootstrap_lower_quantile_point_estimate: Number = np.mean(bootstrap_lower_quantiles)
+    bootstrap_lower_quantile_standard_error: Number = np.std(bootstrap_lower_quantiles)
```
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/helpers/util.py:377]: @NathanFarmer Would it be more accurate to call this variable `bootstrap_lower_quantile_standard_deviation` rather than `bootstrap_lower_quantile_standard_error`?  Thanks!

File: great_expectations/rule_based_profiler/helpers/util.py [hunk 14]
```diff
@@ -297,23 +358,148 @@ def compute_bootstrap_quantiles(
...
+    bootstrap_lower_quantile_bias: Number = (
+        bootstrap_lower_quantile_point_estimate - sample_lower_quantile
+    )
+
+    # Bias / Standard Error > 0.25 is a rule of thumb for when to apply bias correction.
```
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/helpers/util.py:382]: @NathanFarmer Is this standard error or standard deviation?

File: great_expectations/rule_based_profiler/helpers/util.py [hunk 15]
```diff
@@ -297,23 +358,148 @@ def compute_bootstrap_quantiles(
...
+        axis=1,
     )
-    upper_quantile = np.mean(
-        np.quantile(
-            bootstraps,
-            q=1.0 - (false_positive_rate / 2),
-            axis=1,
-            interpolation="linear",  # can be omitted ("linear" is default)
+    bootstrap_upper_quantile_point_estimate: Number = np.mean(bootstrap_upper_quantiles)
+    bootstrap_upper_quantile_standard_error: Number = np.std(bootstrap_upper_quantiles)
```
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/helpers/util.py:402]: @NathanFarmer Standard error or standard deviation here?  Thanks.

File: great_expectations/rule_based_profiler/helpers/util.py [hunk 16]
```diff
@@ -297,23 +358,148 @@ def compute_bootstrap_quantiles(
...
+
+    # The idea that we can take the midpoint of the confidence interval is based on the fact that we think the
+    # confidence interval was built from a symmetrical distribution. We think the distribution is normal due to
+    # the implications of the Central Limit Theorem (CLT) (https://en.wikipedia.org/wiki/Central_limit_theorem) on
+    # the bootstrap samples.
+    # Unfortunately, the assumption that the CLT applies, does not hold in all cases. The bias-corrected and accelerated
+    # (BCa) confidence interval computed using scipy.stats.bootstrap attempts to compute the "acceleration" as a
+    # correction, because the standard normal approximation (CLT) assumes that the standard error of the bootstrap
+    # quantiles (theta-hat) is the same for all parameters (theta). The acceleration (which is the rate of change of
+    # the standard error of the quantile point estimate) is not a perfect correction, and therefore the assumption that
```
Comment (alexsherstinsky) [inline great_expectations/rule_based_profiler/helpers/util.py:481]: @NathanFarmer So you basically implemented a custom version of the "BCa" algorithm?

File: tests/rule_based_profiler/parameter_builder/conftest.py [hunk 17]
```diff
@@ -0,0 +1,184 @@
+from numbers import Number
+from typing import Dict, Optional
+
+import numpy as np
+import pandas as pd
+import pytest
+import scipy.stats as stats
+
+
+def _generate_distribution_parameters(
```
Comment (alexsherstinsky) [inline tests/rule_based_profiler/parameter_builder/conftest.py]: @NathanFarmer How come this is a private method?  Any harm making it public?  Thanks!

File: tests/rule_based_profiler/parameter_builder/conftest.py [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+            "shape": exponential_shape,
+            "lower_bound": exponential_lower_bound,
+            "scale": exponential_scale,
+            "lower_quantile": exponential_lower_quantile,
+            "upper_quantile": exponential_upper_quantile,
+        },
+    }
+
+
+def _generate_distribution_samples(
```
Comment (alexsherstinsky) [inline tests/rule_based_profiler/parameter_builder/conftest.py]: @NathanFarmer Any harm making this method public?  Thanks.

File: tests/rule_based_profiler/parameter_builder/test_parameter_computations.py [hunk 19]
```diff
@@ -1,57 +1,341 @@
...
-                    stats.norm.rvs(4000, 500, size=size // 2),
-                    stats.norm.rvs(6000, 500, size=size // 2),
-                ]
+# Measure of "closeness" between "actual" and "desired" is computed as: atol + rtol * abs(desired)
+# (see "https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html" for details).
+RTOL: float = 1.0e-7
+ATOL: float = 1.0e-2
+
+
+def _compute_quantile_root_mean_squared_error_of_bootstrap(
```
Comment (alexsherstinsky) [inline tests/rule_based_profiler/parameter_builder/test_parameter_computations.py]: @NathanFarmer Any harm making this method public?  Thanks!

**Review by alexsherstinsky — APPROVED** [review]
@NathanFarmer Approving provided you answer some of my questions and act upon them in whichever way you deem appropriate.  This was amazing -- thanks a lot!!!

File: great_expectations/rule_based_profiler/helpers/util.py (hunk 13 above)
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/helpers/util.py:377]: @alexsherstinsky the bootstrap sample standard deviation is referred to as the standard error in this context. See:
```
Efron, B., & Tibshirani, R. J. (1993). Standard errors and estimated standard errors. An Introduction to the Bootstrap (pp. 40).
        Springer Science and Business Media Dordrecht. DOI 10.1007/978-1-4899-4541-9
```

File: great_expectations/rule_based_profiler/helpers/util.py (hunk 14 above)
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/helpers/util.py:382]: @alexsherstinsky the bootstrap sample standard deviation is referred to as the standard error in this context. See:
```
Efron, B., & Tibshirani, R. J. (1993). Standard errors and estimated standard errors. An Introduction to the Bootstrap (pp. 40).
        Springer Science and Business Media Dordrecht. DOI 10.1007/978-1-4899-4541-9
```

File: great_expectations/rule_based_profiler/helpers/util.py (hunk 15 above)
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/helpers/util.py:402]: @alexsherstinsky the bootstrap sample standard deviation is referred to as the standard error in this context. See:
```
Efron, B., & Tibshirani, R. J. (1993). Standard errors and estimated standard errors. An Introduction to the Bootstrap (pp. 40).
        Springer Science and Business Media Dordrecht. DOI 10.1007/978-1-4899-4541-9
```

File: great_expectations/rule_based_profiler/helpers/util.py (hunk 16 above)
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/helpers/util.py:481]: @alexsherstinsky The BCa method is only used to generate confidence intervals. I implemented a custom algorithm that improved on your "mean of the bootstraps" method by correcting for bias if bias was detected above a rule of thumb threshold.

File: tests/rule_based_profiler/parameter_builder/conftest.py (hunk 17 above)
Comment (NathanFarmer) [reply tests/rule_based_profiler/parameter_builder/conftest.py]: @alexsherstinsky I see no harm in making these methods public, but for what purpose?

File: tests/rule_based_profiler/parameter_builder/conftest.py (hunk 18 above)
Comment (NathanFarmer) [reply tests/rule_based_profiler/parameter_builder/conftest.py]: @alexsherstinsky I see no harm in making these methods public, but for what purpose?

File: tests/rule_based_profiler/parameter_builder/test_parameter_computations.py (hunk 19 above)
Comment (NathanFarmer) [reply tests/rule_based_profiler/parameter_builder/test_parameter_computations.py]: @alexsherstinsky I see no harm in making these methods public, but for what purpose?

Comment (alexsherstinsky) [reply tests/rule_based_profiler/parameter_builder/test_parameter_computations.py]: @NathanFarmer So that they can be used in other classes/methods...

File: tests/rule_based_profiler/parameter_builder/conftest.py (hunk 18 above)
Comment (alexsherstinsky) [reply tests/rule_based_profiler/parameter_builder/conftest.py]: @NathanFarmer For use by other classes/method -- seems like useful samples generator.  Thanks!

File: tests/rule_based_profiler/parameter_builder/conftest.py (hunk 17 above)
Comment (alexsherstinsky) [reply tests/rule_based_profiler/parameter_builder/conftest.py]: @NathanFarmer  For use by other classes/method -- seems like a very useful samples generator. Thanks!

File: great_expectations/rule_based_profiler/helpers/util.py (hunk 16 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/helpers/util.py:481]: @NathanFarmer So let me understand this -- our winning algorithm DOES NOT use the `BCa` method?  And the winning algorithm is just an improvement on what we had all along?  Please let me know.  Thank you.

Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/helpers/util.py:481]: @alexsherstinsky that is correct. The winning algorithm and an example can be found here:
```
Efron, B., & Tibshirani, R. J. (1993). Estimates of bias. An Introduction to the Bootstrap (pp. 124-130).
    Springer Science and Business Media Dordrecht. DOI 10.1007/978-1-4899-4541-9
```
